src/main/java/org/prebid/server/bidder/alkimi/AlkimiBidder.java

```diff
@@ -0,0 +1,176 @@
+                .collect(Collectors.toList());
+    }
+
+    private static BidType getBidType(String impId, List<Imp> imps) {
+        BidType bidType = BidType.banner;
+        for (Imp imp : imps) {
+            if (imp.getId().equals(impId)) {
+                if (imp.getBanner() != null) {
+                    return bidType;
+                } else if (imp.getVideo() != null) {
+                    bidType = BidType.video;
+                } else if (imp.getXNative() != null) {
+                    bidType = BidType.xNative;
+                } else if (imp.getAudio() != null) {
+                    bidType = BidType.audio;
+                }
+            }
+        }
+        return bidType;
+    }
+
+    private ExtImpAlkimi parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), ALKIMI_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException("Missing bidder ext in impression with id: " + imp.getId());
+        }
+    }
+
+    private ObjectNode writeImpExt(ExtImpAlkimi extImpAlkimi) {
+        try {
+            return mapper.mapper().convertValue(Map.of("bidder", extImpAlkimi), ObjectNode.class);
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException("Error while writing Ext in Alkimi Bidder");
+        }
+    }
+
+    private Imp updateImp(Imp imp, ExtImpAlkimi ext) {
+        BigDecimal bidFloor = ext.getBidFloor();
+        Integer position = ext.getPos();
+
+        Imp.ImpBuilder impBuilder = imp.toBuilder();
+        ExtImpAlkimi.ExtImpAlkimiBuilder extBuilder = ext.toBuilder();
+        impBuilder.bidfloor(bidFloor);
+
+        if (imp.getBanner() != null) {
+            final Banner banner = imp.getBanner();
+            if (CollectionUtils.isNotEmpty(banner.getFormat())) {
+                final Format firstFormat = banner.getFormat().get(0);
+                impBuilder
+                        .banner(banner.toBuilder()
+                                .w(firstFormat.getW())
+                                .h(firstFormat.getH())
+                                .pos(position)
+                                .build());
+                extBuilder.width(firstFormat.getW());
+                extBuilder.height(firstFormat.getH());
+                extBuilder.impMediaType(TYPE_BANNER);
+            }
+        }
+
+        if (imp.getVideo() != null) {
+            Video video = imp.getVideo();
+            impBuilder
+                    .video(video.toBuilder()
+                            .pos(position)
+                            .build());
+            extBuilder.width(video.getW());
+            extBuilder.height(video.getH());
+            extBuilder.impMediaType(TYPE_VIDEO);
+        }
+
+        impBuilder.ext(writeImpExt(extBuilder.build()));
+        return impBuilder.build();
+    }
```

Comment:
Can be divided into:
```
private Imp updateImp(Imp imp, ExtImpAlkimi extImpAlkimi) {
        final Integer position = extImpAlkimi.getPos();
        final Banner updatedBanner = updateBanner(imp.getBanner(), position);
        final Video updatedVideo = updateVideo(imp.getVideo(), position);

        return imp.toBuilder()
                .bidfloor(extImpAlkimi.getBidFloor())
                .banner(updatedBanner)
                .video(updatedVideo)
                .ext(makeImpExt(updatedBanner, updatedVideo, extImpAlkimi))
                .build();
    }

    private Banner updateBanner(Banner banner, Integer position) {
        if (banner == null || CollectionUtils.isEmpty(banner.getFormat())) {
            return banner;
        }

        final Format firstFormat = banner.getFormat().get(0);
        return banner.toBuilder()
                .w(firstFormat.getW())
                .h(firstFormat.getH())
                .pos(position)
                .build();
    }

    private Video updateVideo(Video video, Integer position) {
        return video != null ? video.toBuilder().pos(position).build() : null;
    }

    private ObjectNode makeImpExt(Banner banner, Video video, ExtImpAlkimi extImpAlkimi) {
        final ExtImpAlkimi.ExtImpAlkimiBuilder extBuilder = extImpAlkimi.toBuilder();

        if (banner != null) {
            extBuilder.width(banner.getW())
                    .height(banner.getH())
                    .impMediaType(TYPE_BANNER);
        }

        if (video != null) {
            extBuilder.width(video.getW());
            extBuilder.height(video.getH());
            extBuilder.impMediaType(TYPE_VIDEO);
        }

        return mapper.mapper().valueToTree(ExtPrebid.of(null, extBuilder.build()));
    }
```

This way it will be easier to read and modify.